src/vllm_tgis_adapter/grpc/adapters.py

```diff
@@ -0,0 +1,142 @@
+
+VALID_ADAPTER_ID_PATTERN = re.compile("[/\\w\\-]+")
+
+
+@dataclasses.dataclass
+class AdapterMetadata:
+    unique_id: int  # Unique integer for vllm to identify the adapter
+    adapter_type: str  # The string name of the peft adapter type, e.g. LORA
+    full_path: str
+
+
+@dataclasses.dataclass
+class AdapterStore:
+    cache_path: str  # Path to local store of adapters to load from
+    adapters: dict[str, AdapterMetadata]
+    next_unique_id: int = 1
+
+
+async def validate_adapters(
+    request: SingleGenerationRequest | BatchedGenerationRequest,
+    adapter_store: AdapterStore | None,
+) -> dict[str, LoRARequest]:
+    """Validate the adapters.
+
+    Takes the adapter name from the request and constructs a valid
+    engine request if one is set. Raises if the requested adapter
+    does not exist or adapter type is unsupported
+
+    Returns the kwarg dictionary to add to an engine.generate() call.
+    """
+    global global_thread_pool  # noqa: PLW0603
+    adapter_id = request.adapter_id
+
+    if adapter_id and not adapter_store:
+        TGISValidationError.AdaptersDisabled.error()
+
+    if not adapter_id or not adapter_store:
+        return {}
+
+    # If not already cached, we need to validate that files exist and
+    # grab the type out of the adapter_config.json file
+    if (adapter_metadata := adapter_store.adapters.get(adapter_id)) is None:
+        _reject_bad_adapter_id(adapter_id)
+        local_adapter_path = Path(adapter_store.cache_path) / adapter_id
+
+        loop = asyncio.get_running_loop()
+        if global_thread_pool is None:
+            global_thread_pool = concurrent.futures.ThreadPoolExecutor(max_workers=2)
+
+        adapter_type = await loop.run_in_executor(
+            global_thread_pool,
+            _get_adapter_type_from_file,
+            adapter_id,
+            local_adapter_path,
+        )
+
+        # Add to cache
+        adapter_metadata = AdapterMetadata(
+            unique_id=adapter_store.next_unique_id,
+            adapter_type=adapter_type,
+            full_path=local_adapter_path,
+        )
+        adapter_store.adapters[adapter_id] = adapter_metadata
+
+    # Build the proper vllm request object
+    if adapter_metadata.adapter_type == "LORA":
+        lora_request = LoRARequest(
+            lora_name=adapter_id,
+            lora_int_id=adapter_metadata.unique_id,
+            lora_local_path=adapter_metadata.full_path,
+        )
+        return {"lora_request": lora_request}
+
+    # All other types unsupported
+    TGISValidationError.AdapterUnsupported.error(adapter_metadata.adapter_type)  # noqa: RET503
+
+
+def _get_adapter_type_from_file(adapter_id: str, adapter_path: str) -> str:
+    """Get adapter from file.
+
+    Performs all the filesystem access required to deduce the type
+    of the adapter. It's run in a separate thread pool executor so that file
+    access does not block the main event loop.
+    """
+    if not Path.exists()(adapter_path):
```

Comment:
This looks wrong:

```suggestion
    not Path(adapter_path).exists()
```

possibly even using `is_file()`